bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/hbasesnapshots/ImportHBaseSnapshotJob.java

```diff
@@ -0,0 +1,286 @@
+  public int run(String[] args) throws Exception {
+    Configuration conf = HBaseConfiguration.create(getConf());
+    String[] otherArgs = new GenericOptionsParser(conf, args).getRemainingArgs();
+
+    if (otherArgs.length < 4) {
+      usage("Wrong number of arguments: " + otherArgs.length);
+      return -1;
+    }
+    if (conf.get(BigtableOptionsFactory.PROJECT_ID_KEY) == null) {
+      usage("Must specify the property " + BigtableOptionsFactory.PROJECT_ID_KEY);
+      return -1;
+    }
+    if (conf.get(BigtableOptionsFactory.INSTANCE_ID_KEY) == null) {
+      usage("Must specify the property " + BigtableOptionsFactory.INSTANCE_ID_KEY);
+      return -1;
+    }
+
+    setConfFromArgs(conf, otherArgs);
+
+    LOG.info(
+        "Using Arguments: "
+            + "\n"
+            + "snapshotName = "
+            + conf.get(SNAPSHOTNAME_KEY)
+            + "\n"
+            + "bigtableName = "
+            + conf.get(TableOutputFormat.OUTPUT_TABLE)
+            + "\n"
+            + "snapshotDir = "
+            + conf.get(HConstants.HBASE_DIR)
+            + "\n"
+            + "restorePath = "
+            + conf.get(SNAPSHOT_RESTOREDIR_KEY)
+            + "\n"
+            + CommonConfigurationKeysPublic.FS_DEFAULT_NAME_KEY
+            + " = "
+            + conf.get(CommonConfigurationKeysPublic.FS_DEFAULT_NAME_KEY)
+            + "\n");
+
+    Job job = createSubmittableJob(conf);
+    int successStatus = (job.waitForCompletion(true) ? 0 : 1);
+
+    Counters counters = job.getCounters();
+    long numRows = counters.findCounter(ScanCounter.NUM_ROWS).getValue();
+    long numCells = counters.findCounter(ScanCounter.NUM_CELLS).getValue();
+
+    System.out.println(("############# Num of ROWS imported= " + numRows));
+    System.out.println(("############# Num of CELLS imported= " + numCells));
+
+    return successStatus;
+  }
+
+  protected static void setConfFromArgs(Configuration conf, String[] args) {
+    // set arguments in configuration for ease of use
+    conf.set(SNAPSHOTNAME_KEY, args[0]);
+    conf.set(TABLENAME_KEY, args[1]);
+    conf.set(HConstants.HBASE_DIR, args[2]);
+    conf.set(SNAPSHOT_RESTOREDIR_KEY, args[3]);
+
+    // set default fs
+    conf.set(CommonConfigurationKeysPublic.FS_DEFAULT_NAME_KEY, conf.get(HConstants.HBASE_DIR));
+
+    // default job name
+    conf.setIfUnset(
+        IMPORT_SNAPSHOT_JOBNAME_KEY,
+        ImportHBaseSnapshotJob.class.getName()
+            + " - from snapshot "
+            + conf.get(SNAPSHOTNAME_KEY)
+            + " to "
+            + conf.get(TABLENAME_KEY));
+
+    // implicit bigtable configs
+    conf.setIfUnset(
+        ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL,
+        BigtableConfiguration.getConnectionClass().getName());
+    conf.setBooleanIfUnset(BigtableOptionsFactory.BIGTABLE_USE_TIMEOUTS_KEY, true);
+    conf.setIfUnset(
+        BigtableOptionsFactory.BIGTABLE_MUTATE_RPC_TIMEOUT_MS_KEY,
+        String.valueOf(BIGTABLE_MUTATE_RPC_TIMEOUT_MS_DEFAULT));
```

Comment:
Is there a way to initialize Bigtable configuration that creates all the default values? If we were to add another config later, how can we ensure that its initialized?

Reply:
most default values should be initialized in `BigtableOptionsFactory` and a few other classes, but this is a simple client override intended to show users how to explicitly set the timeout for batch mutations (and increase it if they timeout). i would assume that other configs added later should be verified by the user, but these shouldn't have a unit test as that would fail if these are adjusted.